proxy/src/main.rs

```diff
@@ -66,39 +67,51 @@ async fn main() -> anyhow::Result<()> {
     let proxy_address: SocketAddr = args.get_one::<String>("proxy").unwrap().parse()?;
     info!("Starting proxy on {proxy_address}");
     let proxy_listener = TcpListener::bind(proxy_address).await?;
+    let cancellation_token = CancellationToken::new();
 
-    let mut tasks = vec![
-        tokio::spawn(handle_signals()),
-        tokio::spawn(http::server::task_main(http_listener)),
-        tokio::spawn(proxy::task_main(config, proxy_listener)),
-        tokio::spawn(console::mgmt::task_main(mgmt_listener)),
-    ];
+    let mut client_tasks = vec![tokio::spawn(proxy::task_main(
+        config,
+        proxy_listener,
+        cancellation_token.clone(),
+    ))];
 
     if let Some(wss_address) = args.get_one::<String>("wss") {
         let wss_address: SocketAddr = wss_address.parse()?;
         info!("Starting wss on {wss_address}");
         let wss_listener = TcpListener::bind(wss_address).await?;
 
-        tasks.push(tokio::spawn(http::websocket::task_main(
+        client_tasks.push(tokio::spawn(http::websocket::task_main(
             config,
             wss_listener,
+            cancellation_token.clone(),
         )));
     }
 
+    let mut tasks = vec![
+        tokio::spawn(handle_signals(cancellation_token)),
+        tokio::spawn(http::server::task_main(http_listener)),
+        tokio::spawn(console::mgmt::task_main(mgmt_listener)),
+    ];
+
     if let Some(metrics_config) = &config.metric_collection {
         tasks.push(tokio::spawn(metrics::task_main(metrics_config)));
     }
 
     // This combinator will block until either all tasks complete or
     // one of them finishes with an error (others will be cancelled).
     let tasks = tasks.into_iter().map(flatten_err);
-    let _: Vec<()> = futures::future::try_join_all(tasks).await?;
-
+    let (tasks_done, abort_handle) =
+        futures::future::abortable(futures::future::try_join_all(tasks));
+    let client_tasks_done =
+        futures::future::try_join_all(client_tasks.into_iter()).then(|_| async move {
+            abort_handle.abort();
+        });
+    let _ = tokio::join!(client_tasks_done, tasks_done);
     Ok(())
```

Comment:
The `then` threw me off a bit.

In the previous design or this new one are we not actually cancelling any of what are now just `tasks`. What we end up aborting is the `try_join_all` of `Vec<tokio::task::JoinHandle>`, which is the same as doing nothing, but with an added unhandled `futures::stream::Aborted` `Err` value. Dropping a `tokio::task::JoinHandle` created by `tokio::spawn` **does nothing** (in the docs this is called "detaches the associated task").

So the statements in selection are equivalent to:

```suggestion
    futures::future::try_join_all(client_tasks.into_iter()).await
```

If we really wanted do de-schedule (or "abort") the `tasks`, then using `JoinSet` for the three would be the easiest way, because all unjoined tasks are by default de-scheduled on drop. I think this would be unnecessary for this PR which just adds graceful stop for the proxy connections. As in we don't need to stop responding via http for example, at least I am unaware of there being any state we should toggle on graceful stop.

Now I noticed the comment:

```rust
// This combinator will block until either all tasks complete or
// one of them finishes with an error (others will be cancelled).
```

I think my suggestion would achieve this because implicitly the rest of the `tokio::spawn`'ed tasks will be dropped when the `async fn main` returns due to dropping of the runtime.

I think there was some comment about a shutdown order being important as opposed your earlier version using `std::process::exit(0)` so I might need to update this comment.

Reply:
> I think there was some comment about a shutdown order being important as opposed your earlier version using `std::process::exit(0)` so I might need to update this comment.

The comment was <https://github.com/neondatabase/neon/pull/3764#discussion_r1132493412>.

Noticed an error with my suggestion. I stand by that it is indeed a minimization of your latest changes, but I don't think it achieves what the original implementation achieved; now an error from any of the `tasks` would go unnoticed.

Easiest way would be to grow the "region of graceful stop" to the two tasks as well (mgmt handler, mgmt http api), then a single `try_join_all` (the original version) would work, as in "early exit on every `Err` but continue until all have completed `Ok(())`".

It might be worth-while to keep the signal handlers task out of this, to keep them just a separate `tokio::spawn`'ed task, because it's unsure if one really wants a signal handling task to ever complete with `Ok(())`.

Reply:
Oki, I had misunderstood. We went over this on a call, and ended up with a simpler `tokio::select` race which handles:
- any of the "forever tasks" (in `tasks`) returning Err
- all of the "graceful shutdowning tasks (in `client_tasks`) returning Ok or Err

We couldn't use `tokio::try_join!` because the forever running tasks never fail, and also this way we don't need to have an "Ok" `Err` type in addition to `anyhow::Result<()>`.